### PR TITLE
feat(publish): Use temporary dist-tags by default

### PR DESCRIPTION
**Pull request by evocateur**
## Description

The previous default was _not_ to use temporary dist-tags during the `lerna publish` process, and thus a partially successful publish (due to authentication errors and the like) would result in a broken `latest` dist-tag.

Pass `--no-temp-tag` to restore the previous, slightly more dangerous behavior.

## Motivation and Context

#1862 should be merged first.

Refs #1767

## How Has This Been Tested?

moar tests

## Types of changes
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by evocateur** [comment]
The primarily interesting commit is https://github.com/lerna/lerna/pull/1863/commits/4eff8d2280c39034b988bcccc8365e99cc97b02e, which does not include the changes from the "upstream" PR.

File: commands/publish/README.md [hunk 1]
```diff
@@ -108,6 +108,13 @@ To avoid this, pass `--no-git-reset`. This can be especially useful when used as
 lerna publish --no-git-reset
 ```
 
+### `--no-temp-tag`
+
+By default, `lerna publish` first publishes all changed packages to a temporary dist-tag (`lerna-temp`)
+and then after all are successful moves the new version(s) to the dist-tag configured by [`--dist-tag`](#--dist-tag-tag) (default `latest`).
+
+If you like to live dangerously, pass `--no-temp-tag` (or set `command.publish.tempTag` to `false` in lerna.json) to disable this behavior.
```
Comment (ThisIsMissEm) [inline commands/publish/README.md]: Perhaps we need to explain why this is dangerous?

File: commands/publish/command.js [hunk 2]
```diff
@@ -40,18 +40,15 @@ exports.builder = yargs => {
...
     "git-reset": {
       describe: "Reset the git working tree.\nPass --no-git-reset to disable.",
       type: "boolean",
       defaultDescription: "true",
     },
     "temp-tag": {
-      describe: "Create a temporary tag while publishing.",
+      describe: "Create a temporary dist-tag while publishing.\nPass --no-temp-tag to disable.",
       type: "boolean",
+      defaultDescription: "true",
```
Comment (ThisIsMissEm) [inline commands/publish/command.js]: How does this change affect the `lerna help` output? Should it actually default to `no-temp-tag` instead of `temp-tag`?

Comment (ThisIsMissEm) [reply commands/publish/command.js]: Also, isn't this capable of using a different string here?

File: commands/publish/README.md (hunk 1 above)
Comment (evocateur) [reply commands/publish/README.md]: Yes, that's a better idea.

File: commands/publish/__tests__/get-unpublished-packages.test.js [hunk 3]
```diff
@@ -27,8 +33,38 @@ getPackument.mockImplementation(async pkg => {
...
-  const project = new Project(cwd);
+  const packages = await getPackages(cwd);
+  const packageGraph = new PackageGraph(packages);
+
+  const opts = new Map();
+  const pkgs = await getUnpublishedPackages(packageGraph, opts);
 
-  const pkgs = await getUnpublishedPackages(project, {});
-  expect(pkgs.map(p => p.name)).toEqual(["package-1", "package-3", "package-4", "package-5"]);
+  expect(pkgs).toMatchInlineSnapshot(`
```
Comment (evocateur) [inline commands/publish/__tests__/get-unpublished-packages.test.js]: These changes are just fallout from unrelated refactoring (prompted by removal of a dependency I thought was clear). In the end, I didn't make the effort to separate it.

File: commands/publish/__tests__/publish-canary-no-git-reset.test.js [hunk 4]
```diff
@@ -73,24 +70,5 @@ test("publish --canary --no-git-reset", async () => {
...
-  expect(writePkg.updatedVersions()).toMatchInlineSnapshot(`
-Object {
-  "package-1": 1.0.1-alpha.0+SHA,
-  "package-2": 1.0.1-alpha.0+SHA,
-  "package-3": 1.0.1-alpha.0+SHA,
-  "package-4": 1.0.1-alpha.0+SHA,
-}
-`);
-  expect(gitCheckout).toHaveBeenCalledTimes(0);
+  expect(gitCheckout).not.toHaveBeenCalled();
```
Comment (evocateur) [inline commands/publish/__tests__/publish-canary-no-git-reset.test.js]: The majority of these assertions were removed because they were completely redundant and obscured the actual point of the test case.

File: commands/publish/__tests__/publish-canary.test.js [hunk 5]
```diff
@@ -14,7 +14,7 @@ const childProcess = require("@lerna/child-process");
 
 // mocked modules
 const writePkg = require("write-pkg");
-const npmPublish = require("@lerna/npm-publish");
+const npmDistTag = require("@lerna/npm-dist-tag");
```
Comment (evocateur) [inline commands/publish/__tests__/publish-canary.test.js]: Most assertions on the final dist-tag now need to use the `@lerna/npm-dist-tag` mock, which explains all the new stuff in the mock helper. (`@lerna/npm-publish` is almost always called with `lerna-temp` dist-tag, which isn't very useful as an assertion)

File: commands/publish/__tests__/publish-command.test.js [hunk 6]
```diff
@@ -259,11 +256,10 @@ Map {
...
     it("publishes unpublished packages", async () => {
       const testDir = await initFixture("normal");
-      const project = new Project(testDir);
 
-      getUnpublishedPackages.mockImplementationOnce(async () => {
-        const pkgs = await project.getPackages();
-        return pkgs.slice(1, 3);
+      getUnpublishedPackages.mockImplementationOnce(packageGraph => {
+        const pkgs = packageGraph.rawPackageList.slice(1, 3);
+        return pkgs.map(pkg => packageGraph.get(pkg.name));
```
Comment (evocateur) [inline commands/publish/__tests__/publish-command.test.js]: another mostly-unrelated change due to refactoring `getUnpublishedPackages()`. it definitely exposes that my mock isn't abstract enough, and should probably be a network mock instead...

File: commands/publish/command.js (hunk 2 above)
Comment (evocateur) [reply commands/publish/command.js]: We're using yargs' "auto-negation" feature here (any boolean option prefixed with `no-` is set to `false`), so it is a bit awkward, I agree. We can't use `"no-temp-tag"` as the key, otherwise it would need to be passed as `--no-no-temp-tag`.

(I could be wrong, it's been awhile since I tried to make this clearer. It could also require upstream changes in yargs)

File: commands/publish/lib/get-unpublished-packages.js [hunk 7]
```diff
@@ -1,33 +1,30 @@
 "use strict";
 
 const log = require("libnpm/log");
-const pReduce = require("p-reduce");
+const pMap = require("p-map");
```
Comment (evocateur) [inline commands/publish/lib/get-unpublished-packages.js]: I removed `p-reduce` from the main command script, forgot it was here, and then realized it could be implemented with more concurrency with `p-map`. (`p-reduce` executes serially)

File: integration/lerna-publish-canary.test.js [hunk 8]
```diff
@@ -22,7 +22,7 @@ const env = {
 
 test("lerna publish --canary uses default prerelease id", async () => {
   const { cwd } = await cloneFixture("normal");
-  const args = ["publish", "--canary", "--yes"];
+  const args = ["publish", "--canary", "--yes", "--no-temp-tag"];
```
Comment (evocateur) [inline integration/lerna-publish-canary.test.js]: I had to add `--no-temp-tag` to all the end-to-end tests because we don't have a good solution for local testing with an _actual_ registry, and none of the test packages are actually "live" on the public registry.

File: commands/publish/__tests__/get-unpublished-packages.test.js (hunk 3 above)
Comment (evocateur) [reply commands/publish/__tests__/get-unpublished-packages.test.js]: Actually, I think I _will_ move this commit out of this PR. It's very noisy.

**Comment by ThisIsMissEm** [comment]
Hmm, okay. That is indeed a little bit awkward.

**Comment by evocateur** [comment]
@ThisIsMissEm I think I figured out a pattern that makes the inverted booleans more ergonomic (documented as `--no-<thing>` in the `-h` output, even).

For example, `--no-progress` is the only option that makes sense, as [`--progress` is implied](https://github.com/lerna/lerna/blob/c79a82722095607940749f5c6fa1cdd90c56a437/core/global-options/index.js#L23-L27) (where progress is even used, getting kinda itchy in the "global" options). This change makes `--no-progress` output in the `--help`, while keeping the shape of the resulting `argv` object (mostly) consistent:
```patch
+    "no-progress": {
+      describe: "Disable progress bars. (Always off in CI)",
       type: "boolean",
     },
     progress: {
-      defaultDescription: "true",
-      describe: "Enable progress bars. (Always off in CI)\nPass --no-progress to disable.",
+      // proxy for --no-progress
+      hidden: true,
+      type: "boolean",
+    },
```

It does technically add new `noProgress` keys, _and_ can understand `--no-no-progress` flags, but I think we can ignore those, in practice. Dunno if I should port this upstream to yargs in a cleaner way, but figure this kinda-hacky solution is at least very... declarative? If a bit mysterious...

There's a bunch of other flag deprecations I want to remove in the semver major, so I'll be doing that soon enough...

**Comment by evocateur** [comment]
this works globally across all commands:
```diff
diff --git a/core/lerna/index.js b/core/lerna/index.js
index a46caba..dbfd308 100644
--- a/core/lerna/index.js
+++ b/core/lerna/index.js
@@ -26,7 +26,7 @@ function main(argv) {
     lernaVersion: pkg.version,
   };
 
-  return cli()
+  const parser = cli()
     .command(addCmd)
     .command(bootstrapCmd)
     .command(changedCmd)
@@ -40,6 +40,23 @@ function main(argv) {
     .command(listCmd)
     .command(publishCmd)
     .command(runCmd)
-    .command(versionCmd)
-    .parse(argv, context);
+    .command(versionCmd);
+
+  invertBooleanOptions(parser);
+
+  return parser.parse(argv, context);
+}
+
+function invertBooleanOptions(parser) {
+  const opts = parser.getOptions();
+
+  for (const key of Object.keys(opts.key)) {
+    if (key.startsWith("no-")) {
+      // --no-foo => inverted --foo
+      parser.option(key.substring(3), {
+        hidden: true,
+        type: "boolean",
+      });
+    }
+  }
 }
```

✨

**Comment by JamesHenry** [comment]
I will close this one because of its age, but we will look to weave it into the recoverable publish work as part of v7
